[liblsquic] Add new port

rpavlik opened the pull request:
Adds a port for lsquic, an implementation of several versions of QUIC and HTTP/3.

The thing I have the biggest question about is the submodules: I tried to follow the technique I saw in a few other ports, but not sure if it's the preferred way.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
I have tested Linux and Windows, have not updated the CI baseline.

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
Yes - patches have already been upstreamed and dropped.

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
yes

rpavlik committed:
[liblsquic] Add new port.

JackBoosY commented on ports/liblsquic/export_targets.patch:
```diff
@@ -0,0 +1,22 @@
+     ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+ )
+ 
++install(
```
Why the upstream doesn't contain the install function although the target was exported? Can you please also make a PR to the upstream?

rpavlik replied:
Yes I have (or at least I put it on a branch, don't remember if I actually submitted it)

rpavlik replied:
https://github.com/litespeedtech/lsquic/pull/370 ok now it's actually submitted.

rpavlik commented:
> You have modified or added at least one vcpkg.json where a "license" field is missing.

I can add this

rpavlik commented:
OK, I have added the license field, removed the unneeded patch, and the install patch has been submitted upstream. Not sure if there's anything I need to do to explicit update the CI baseline, since it appears to build everywhere just fine.

rpavlik commented:
Oh, hmm, I missed this detail in the maintainer guide at first: "Libraries that don't provide a .def file and do not use __declspec() declarations simply do not support shared builds for Windows and should be marked as such with vcpkg_check_linkage(ONLY_STATIC_LIBRARY)."  I think it only provides a C api, do I still need to add this line? It's been apparently working OK with shared builds on Windows here...

rpavlik commented:
OK I added `vcpkg_check_linkage(ONLY_STATIC_LIBRARY)` on Windows in the portfile.

JackBoosY commented on ports/liblsquic/portfile.cmake:
```diff
@@ -0,0 +1,82 @@
+    file(RENAME "${LSHPACK_SOURCE_PATH}" "${SOURCE_PATH}/src/lshpack")
+endif()
+
+# Configuration
+if(WIN32)
+  # The lib uses CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS, at least until
+  # https://github.com/litespeedtech/lsquic/pull/371 or similar is merged
+  vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
+endif()
```
I'd like you to put this paragraph at the top of this file.

rpavlik replied:
Done

JackBoosY commented on ports/liblsquic/portfile.cmake:
```diff
@@ -0,0 +1,82 @@
+  vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
+endif()
+vcpkg_find_acquire_program(PERL)
+get_filename_component(PERL_EXE_PATH ${PERL} DIRECTORY)
+vcpkg_add_to_path(${PERL_EXE_PATH})
```
```suggestion
get_filename_component(PERL_EXE_PATH "${PERL}" DIRECTORY)
vcpkg_add_to_path("${PERL_EXE_PATH}")
```
Please also add double quotes to all absolute paths.

rpavlik commented:
OK, revised as requested.

dg0yt commented on ports/liblsquic/portfile.cmake:
```diff
@@ -0,0 +1,82 @@
+# Configuration
+vcpkg_find_acquire_program(PERL)
+get_filename_component(PERL_EXE_PATH "${PERL}" DIRECTORY)
+vcpkg_add_to_path("${PERL_EXE_PATH}")
```
Is it necessary to add this to `PATH`, or could the location of perl be passed as an input variable via CMake options?

rpavlik replied:
Ah. It would work on Windows, where perl is only used to generate headers. On Linux, it's called from within `print-glibc-version.sh` with using the path.

rpavlik replied:
Would you prefer that I leave this like this, or patch the script they're using on Linux so we can pass an alternate perl location? (or make this line linux only, or replace the one problematic perl use with this https://github.com/litespeedtech/lsquic/pull/374 ) ?

rpavlik replied:
my fix for this was accepted upstream.

BillyONeal commented on ports/liblsquic/vcpkg.json:
```diff
@@ -0,0 +1,23 @@
+  "homepage": "https://github.com/litespeedtech/lsquic",
+  "license": "MIT AND BSD-3-Clause",
+  "dependencies": [
+    "boringssl",
```
Note that this means we are not able to test this port in the vcpkg build lab. (And our boringssl is probably very out of date a lot of the time as a result)

rpavlik replied:
For legal reasons or something? Or just because of the incompatibility with openssl?

Unfortunately this is a hard dependency for actual usage, boringssl has some API entry points that are specific for it. I did find a build option to disable encryption, not sure how useful it is. On the plus side, it's quite happy with the version of boringssl vcpkg has right now!

BillyONeal replied:
Because of the incompatibility with openssl, yes.

rpavlik replied:
that's a bummer... Too bad google didn't namespace boringssl better.

BillyONeal commented:
>The thing I have the biggest question about is the submodules: I tried to follow the technique I saw in a few other ports, but not sure if it's the preferred way.

Looks good to me!

BillyONeal commented:
On x86-windows the build fails with:

```
C:\PROGRA~1\MIB055~1\2022\ENTERP~1\VC\Tools\MSVC\1431~1.311\bin\Hostx64\x86\cl.exe    -IC:\Dev\vcpkg\buildtrees\liblsquic\x86-windows-dbg\src\liblsquic -IC:\Dev\vcpkg\buildtrees\liblsquic\src\v3.0.4-6f75400bac.clean\src\liblsquic -IC:\Dev\vcpkg\installed\x86-windows\include -IC:\Dev\vcpkg\buildtrees\liblsquic\src\v3.0.4-6f75400bac.clean\include -IC:\Dev\vcpkg\buildtrees\liblsquic\src\v3.0.4-6f75400bac.clean\src\liblsquic\ls-qpack -IC:\Dev\vcpkg\buildtrees\liblsquic\src\v3.0.4-6f75400bac.clean\src\liblsquic\..\lshpack /nologo /DWIN32 /D_WINDOWS /W3 /utf-8 /MP   -DLSQUIC_DEBUG_NEXT_ADV_TICK=1 -DLSQUIC_CONN_STATS=1 /wd4100 /wd4115 /wd4116 /wd4146 /wd4132 /wd4200 /wd4204 /wd4244 /wd4245 /wd4267 /wd4214 /wd4295 /wd4334 /wd4456 /wd4459 /wd4706 /wd4090 /wd4305 -W4 -WX -Zi -DWIN32_LEAN_AND_MEAN -DNOMINMAX -D_CRT_SECURE_NO_WARNINGS -IC:/Dev/vcpkg/buildtrees/liblsquic/src/v3.0.4-6f75400bac.clean/wincompat -Od  -DXXH_HEADER_NAME=\"lsquic_xxhash.h\" -DLSQPACK_ENC_LOGGER_HEADER=\"lsquic_qpack_enc_logger.h\" -DLSQPACK_DEC_LOGGER_HEADER=\"lsquic_qpack_dec_logger.h\" /D_DEBUG /MDd /Z7 /Ob0 /Od /RTC1 /showIncludes /Fosrc\liblsquic\CMakeFiles\lsquic.dir\lsquic_rechist.c.obj /Fdsrc\liblsquic\CMakeFiles\lsquic.dir\lsquic.pdb /FS -c C:\Dev\vcpkg\buildtrees\liblsquic\src\v3.0.4-6f75400bac.clean\src\liblsquic\lsquic_rechist.c
cl : Command line warning D9025 : overriding '/W3' with '/W4'
cl : Command line warning D9025 : overriding '/Zi' with '/Z7'
C:\Dev\vcpkg\buildtrees\liblsquic\src\v3.0.4-6f75400bac.clean\src\liblsquic\lsquic_rechist.c(321): error C2220: the following warning is treated as an error
C:\Dev\vcpkg\buildtrees\liblsquic\src\v3.0.4-6f75400bac.clean\src\liblsquic\lsquic_rechist.c(321): warning C4389: '==': signed/unsigned mismatch
```

Normally we would just remove -WX but it looks like this port seems to intentionally not care about x86, should that be blocked with "supports"?

rpavlik commented:
I'll be honest that I hadn't really tried 32 bit, but I agree its build system is quite opinionated regarding flags, etc. I've got a cleanup submitted but no idea if there's any appetite to merge. I think the software I'm building should work on 32 bit, eventually. Would you prefer that I leave it as-is and just mark it as not ok for x86?

rpavlik commented:
> On x86-windows the build fails with:
> 
> ```
> C:\PROGRA~1\MIB055~1\2022\ENTERP~1\VC\Tools\MSVC\1431~1.311\bin\Hostx64\x86\cl.exe    -IC:\Dev\vcpkg\buildtrees\liblsquic\x86-windows-dbg\src\liblsquic -IC:\Dev\vcpkg\buildtrees\liblsquic\src\v3.0.4-6f75400bac.clean\src\liblsquic -IC:\Dev\vcpkg\installed\x86-windows\include -IC:\Dev\vcpkg\buildtrees\liblsquic\src\v3.0.4-6f75400bac.clean\include -IC:\Dev\vcpkg\buildtrees\liblsquic\src\v3.0.4-6f75400bac.clean\src\liblsquic\ls-qpack -IC:\Dev\vcpkg\buildtrees\liblsquic\src\v3.0.4-6f75400bac.clean\src\liblsquic\..\lshpack /nologo /DWIN32 /D_WINDOWS /W3 /utf-8 /MP   -DLSQUIC_DEBUG_NEXT_ADV_TICK=1 -DLSQUIC_CONN_STATS=1 /wd4100 /wd4115 /wd4116 /wd4146 /wd4132 /wd4200 /wd4204 /wd4244 /wd4245 /wd4267 /wd4214 /wd4295 /wd4334 /wd4456 /wd4459 /wd4706 /wd4090 /wd4305 -W4 -WX -Zi -DWIN32_LEAN_AND_MEAN -DNOMINMAX -D_CRT_SECURE_NO_WARNINGS -IC:/Dev/vcpkg/buildtrees/liblsquic/src/v3.0.4-6f75400bac.clean/wincompat -Od  -DXXH_HEADER_NAME=\"lsquic_xxhash.h\" -DLSQPACK_ENC_LOGGER_HEADER=\"lsquic_qpack_enc_logger.h\" -DLSQPACK_DEC_LOGGER_HEADER=\"lsquic_qpack_dec_logger.h\" /D_DEBUG /MDd /Z7 /Ob0 /Od /RTC1 /showIncludes /Fosrc\liblsquic\CMakeFiles\lsquic.dir\lsquic_rechist.c.obj /Fdsrc\liblsquic\CMakeFiles\lsquic.dir\lsquic.pdb /FS -c C:\Dev\vcpkg\buildtrees\liblsquic\src\v3.0.4-6f75400bac.clean\src\liblsquic\lsquic_rechist.c
> cl : Command line warning D9025 : overriding '/W3' with '/W4'
> cl : Command line warning D9025 : overriding '/Zi' with '/Z7'
> C:\Dev\vcpkg\buildtrees\liblsquic\src\v3.0.4-6f75400bac.clean\src\liblsquic\lsquic_rechist.c(321): error C2220: the following warning is treated as an error
> C:\Dev\vcpkg\buildtrees\liblsquic\src\v3.0.4-6f75400bac.clean\src\liblsquic\lsquic_rechist.c(321): warning C4389: '==': signed/unsigned mismatch
> ```
> 
> Normally we would just remove -WX but it looks like this port seems to intentionally not care about x86, should that be blocked with "supports"?

OK, I have tried doing this, I can rebase and flatten and regen versions if I get the approval that I did it right. This is my first time upstreaming a package to vcpkg (though I maintain several debian packages) so I'm still unsure of my footing.

(Also, that's totally a weird warning to occur when switching 64 to 32, didn't realize any types changed default sign)

rpavlik commented:
OK, so upstream is merging a bunch of my patches, I'll update this once I hear back about how I limited the platforms, and once they tag their next release.

rpavlik commented:
OK, upstream has released, I've updated accordingly, think this should be good to go once CI passes.

BillyONeal commented:
/azp run

BillyONeal commented:
Did we get to an answer on whether x86 should be supported? I observe that your supports expression currently blocks it

BillyONeal commented:
Reminder to reviewers: CI does not test this change because it depends on boringssl which is blocked in CI.

I was unable to test this myself because boringssl does not build with current Visual Studio:

```
C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\include\openssl/stack.h(428): error C2220: the following warning is treated as an error
C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\include\openssl/stack.h(428): warning C4191: 'type cast': unsafe conversion from 'stack_free_func' to 'stack_OPENSSL_STRING_free_func'
        Calling this function through the result pointer may cause your program to fail
C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\include\openssl/stack.h(428): warning C4191: 'type cast': unsafe conversion from 'stack_copy_func' to 'stack_OPENSSL_STRING_copy_func'
        Calling this function through the result pointer may cause your program to fail
C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\include\openssl/stack.h(428): warning C4191: 'type cast': unsafe conversion from 'stack_cmp_func' to 'stack_OPENSSL_STRING_cmp_func'
        Calling this function through the result pointer may cause your program to fail
C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\include\openssl/stack.h(428): warning C4191: 'type cast': unsafe conversion from 'stack_OPENSSL_STRING_cmp_func' to 'stack_cmp_func'
        Calling this function through the result pointer may cause your program to fail
C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\include\openssl/stack.h(428): warning C4191: 'type cast': unsafe conversion from 'stack_OPENSSL_STRING_free_func' to 'stack_free_func'
        Calling this function through the result pointer may cause your program to fail
C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\include\openssl/stack.h(428): warning C4191: 'type cast': unsafe conversion from 'OPENSSL_STRING (__cdecl *)(OPENSSL_STRING)' to 'stack_copy_func'
        Calling this function through the result pointer may cause your program to fail
C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\include\openssl/stack.h(428): warning C4191: 'type cast': unsafe conversion from 'void (__cdecl *)(OPENSSL_STRING)' to 'stack_free_func'
        Calling this function through the result pointer may cause your program to fail
C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\crypto\fipsmodule\bn/ctx.c(88): warning C4191: 'type cast': unsafe conversion from 'stack_free_func' to 'stack_BIGNUM_free_func'
        Calling this function through the result pointer may cause your program to fail
C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\crypto\fipsmodule\bn/ctx.c(88): warning C4191: 'type cast': unsafe conversion from 'stack_copy_func' to 'stack_BIGNUM_copy_func'
        Calling this function through the result pointer may cause your program to fail
C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\crypto\fipsmodule\bn/ctx.c(88): warning C4191: 'type cast': unsafe conversion from 'stack_cmp_func' to 'stack_BIGNUM_cmp_func'
        Calling this function through the result pointer may cause your program to fail
C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\crypto\fipsmodule\bn/ctx.c(88): warning C4191: 'type cast': unsafe conversion from 'stack_BIGNUM_cmp_func' to 'stack_cmp_func'
        Calling this function through the result pointer may cause your program to fail
C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\crypto\fipsmodule\bn/ctx.c(88): warning C4191: 'type cast': unsafe conversion from 'stack_BIGNUM_free_func' to 'stack_free_func'
        Calling this function through the result pointer may cause your program to fail
C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\crypto\fipsmodule\bn/ctx.c(88): warning C4191: 'type cast': unsafe conversion from 'BIGNUM *(__cdecl *)(BIGNUM *)' to 'stack_copy_func'
        Calling this function through the result pointer may cause your program to fail
C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\crypto\fipsmodule\bn/ctx.c(88): warning C4191: 'type cast': unsafe conversion from 'void (__cdecl *)(BIGNUM *)' to 'stack_free_func'
        Calling this function through the result pointer may cause your program to fail
[39/440] cmd.exe /C "cd . && C:\PROGRA~1\MIB055~1\2022\ENTERP~1\VC\Tools\MSVC\1432~1.313\bin\Hostx64\x64\lib.exe  /machine:x64 /nologo /out:crypto\test\test_support_lib.lib crypto\test\CMakeFiles\test_support_lib.dir\abi_test.cc.obj crypto\test\CMakeFiles\test_support_lib.dir\file_test.cc.obj crypto\test\CMakeFiles\test_support_lib.dir\malloc.cc.obj crypto\test\CMakeFiles\test_support_lib.dir\test_util.cc.obj crypto\test\CMakeFiles\test_support_lib.dir\wycheproof_util.cc.obj  && cd ."
[40/440] C:\PROGRA~1\MIB055~1\2022\ENTERP~1\VC\Tools\MSVC\1432~1.313\bin\Hostx64\x64\cl.exe   /TP -DBORINGSSL_DISPATCH_TEST -DBORINGSSL_IMPLEMENTATION -DNOMINMAX -DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_WARNINGS -D_HAS_EXCEPTIONS=0 -D_STL_EXTRA_DISABLED_WARNINGS="4774 4987" -IC:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\third_party\googletest -IC:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\third_party\googletest\include -utf-8 -Wall -WX  -wd4061 -wd4100 -wd4127 -wd4200 -wd4204 -wd4221 -wd4242 -wd4244 -wd4267 -wd4371 -wd4388 -wd4296 -wd4350 -wd4365 -wd4389 -wd4464 -wd4510 -wd4512 -wd4514 -wd4548 -wd4610 -wd4623 -wd4625 -wd4626 -wd4628 -wd4668 -wd4706 -wd4710 -wd4711 -wd4800 -wd4820 -wd5026 -wd5027 -wd5045  -w44265 /D_DEBUG /MTd /Z7 /Ob0 /Od /RTC1 /showIncludes /FoCMakeFiles\boringssl_gtest.dir\third_party\googletest\src\gtest-all.cc.obj /FdCMakeFiles\boringssl_gtest.dir\boringssl_gtest.pdb /FS -c C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\third_party\googletest\src\gtest-all.cc
Microsoft (R) C/C++ Optimizing Compiler Version 19.32.31329 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.
```

BillyONeal commented:
> Did we get to an answer on whether x86 should be supported? I observe that your supports expression currently blocks it

Adding requires:author-response back over this

rpavlik commented:
re: BoringSSL build failure, I hadn't seen that because I was on a slightly older vs2022, so I assumed that when you said "latest" you meant some internal dogfood edition of MSVC. I have filed an issue upstream, apparently their CI tops out at 2019. https://bugs.chromium.org/p/boringssl/issues/detail?id=495 

Filed issue upstream about 32 bit support, since I am but a humble user/build system contributor of the project. https://github.com/litespeedtech/lsquic/issues/382

rpavlik commented:
> Reminder to reviewers: CI does not test this change because it depends on boringssl which is blocked in CI.
> 
> I was unable to test this myself because boringssl does not build with current Visual Studio:
> 
> ```
> C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\include\openssl/stack.h(428): error C2220: the following warning is treated as an error
> C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\include\openssl/stack.h(428): warning C4191: 'type cast': unsafe conversion from 'stack_free_func' to 'stack_OPENSSL_STRING_free_func'
>         Calling this function through the result pointer may cause your program to fail
> C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\include\openssl/stack.h(428): warning C4191: 'type cast': unsafe conversion from 'stack_copy_func' to 'stack_OPENSSL_STRING_copy_func'
>         Calling this function through the result pointer may cause your program to fail
> C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\include\openssl/stack.h(428): warning C4191: 'type cast': unsafe conversion from 'stack_cmp_func' to 'stack_OPENSSL_STRING_cmp_func'
>         Calling this function through the result pointer may cause your program to fail
> C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\include\openssl/stack.h(428): warning C4191: 'type cast': unsafe conversion from 'stack_OPENSSL_STRING_cmp_func' to 'stack_cmp_func'
>         Calling this function through the result pointer may cause your program to fail
> C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\include\openssl/stack.h(428): warning C4191: 'type cast': unsafe conversion from 'stack_OPENSSL_STRING_free_func' to 'stack_free_func'
>         Calling this function through the result pointer may cause your program to fail
> C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\include\openssl/stack.h(428): warning C4191: 'type cast': unsafe conversion from 'OPENSSL_STRING (__cdecl *)(OPENSSL_STRING)' to 'stack_copy_func'
>         Calling this function through the result pointer may cause your program to fail
> C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\include\openssl/stack.h(428): warning C4191: 'type cast': unsafe conversion from 'void (__cdecl *)(OPENSSL_STRING)' to 'stack_free_func'
>         Calling this function through the result pointer may cause your program to fail
> C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\crypto\fipsmodule\bn/ctx.c(88): warning C4191: 'type cast': unsafe conversion from 'stack_free_func' to 'stack_BIGNUM_free_func'
>         Calling this function through the result pointer may cause your program to fail
> C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\crypto\fipsmodule\bn/ctx.c(88): warning C4191: 'type cast': unsafe conversion from 'stack_copy_func' to 'stack_BIGNUM_copy_func'
>         Calling this function through the result pointer may cause your program to fail
> C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\crypto\fipsmodule\bn/ctx.c(88): warning C4191: 'type cast': unsafe conversion from 'stack_cmp_func' to 'stack_BIGNUM_cmp_func'
>         Calling this function through the result pointer may cause your program to fail
> C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\crypto\fipsmodule\bn/ctx.c(88): warning C4191: 'type cast': unsafe conversion from 'stack_BIGNUM_cmp_func' to 'stack_cmp_func'
>         Calling this function through the result pointer may cause your program to fail
> C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\crypto\fipsmodule\bn/ctx.c(88): warning C4191: 'type cast': unsafe conversion from 'stack_BIGNUM_free_func' to 'stack_free_func'
>         Calling this function through the result pointer may cause your program to fail
> C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\crypto\fipsmodule\bn/ctx.c(88): warning C4191: 'type cast': unsafe conversion from 'BIGNUM *(__cdecl *)(BIGNUM *)' to 'stack_copy_func'
>         Calling this function through the result pointer may cause your program to fail
> C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\crypto\fipsmodule\bn/ctx.c(88): warning C4191: 'type cast': unsafe conversion from 'void (__cdecl *)(BIGNUM *)' to 'stack_free_func'
>         Calling this function through the result pointer may cause your program to fail
> [39/440] cmd.exe /C "cd . && C:\PROGRA~1\MIB055~1\2022\ENTERP~1\VC\Tools\MSVC\1432~1.313\bin\Hostx64\x64\lib.exe  /machine:x64 /nologo /out:crypto\test\test_support_lib.lib crypto\test\CMakeFiles\test_support_lib.dir\abi_test.cc.obj crypto\test\CMakeFiles\test_support_lib.dir\file_test.cc.obj crypto\test\CMakeFiles\test_support_lib.dir\malloc.cc.obj crypto\test\CMakeFiles\test_support_lib.dir\test_util.cc.obj crypto\test\CMakeFiles\test_support_lib.dir\wycheproof_util.cc.obj  && cd ."
> [40/440] C:\PROGRA~1\MIB055~1\2022\ENTERP~1\VC\Tools\MSVC\1432~1.313\bin\Hostx64\x64\cl.exe   /TP -DBORINGSSL_DISPATCH_TEST -DBORINGSSL_IMPLEMENTATION -DNOMINMAX -DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_WARNINGS -D_HAS_EXCEPTIONS=0 -D_STL_EXTRA_DISABLED_WARNINGS="4774 4987" -IC:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\third_party\googletest -IC:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\third_party\googletest\include -utf-8 -Wall -WX  -wd4061 -wd4100 -wd4127 -wd4200 -wd4204 -wd4221 -wd4242 -wd4244 -wd4267 -wd4371 -wd4388 -wd4296 -wd4350 -wd4365 -wd4389 -wd4464 -wd4510 -wd4512 -wd4514 -wd4548 -wd4610 -wd4623 -wd4625 -wd4626 -wd4628 -wd4668 -wd4706 -wd4710 -wd4711 -wd4800 -wd4820 -wd5026 -wd5027 -wd5045  -w44265 /D_DEBUG /MTd /Z7 /Ob0 /Od /RTC1 /showIncludes /FoCMakeFiles\boringssl_gtest.dir\third_party\googletest\src\gtest-all.cc.obj /FdCMakeFiles\boringssl_gtest.dir\boringssl_gtest.pdb /FS -c C:\Dev\vcpkg\buildtrees\boringssl\src\20e1d1a6d6-ca289bd3ff.clean\third_party\googletest\src\gtest-all.cc
> Microsoft (R) C/C++ Optimizing Compiler Version 19.32.31329 for x64
> Copyright (C) Microsoft Corporation.  All rights reserved.
> ```

BoringSSL upstream has merged a VS2022 build fix.

https://boringssl.googlesource.com/boringssl/+/167f1760ddfeaea1ee1a671ca88aafcccfe30ee0

PhoebeHui commented:
@rpavlik

PhoebeHui commented on ports/liblsquic/portfile.cmake:
```diff
@@ -0,0 +1,70 @@
+endif()
+
+# Configuration
+vcpkg_find_acquire_program(PERL)
```
```suggestion
vcpkg_find_acquire_program(PERL)
get_filename_component(PERL_PATH ${PERL} DIRECTORY)
vcpkg_add_to_path(${PERL_PATH})
```

rpavlik replied:
OK, yeah, just have to pass PERL as both PERL and PERL_EXECUTABLE.

PhoebeHui commented on ports/liblsquic/portfile.cmake:
```diff
@@ -0,0 +1,70 @@
+vcpkg_cmake_configure(
+  SOURCE_PATH "${SOURCE_PATH}"
+  OPTIONS
+    "-DPERL=${PERL}"
```
```suggestion
```

dg0yt replied:
I don't think it is to be removed. But the current code uses `find_package(Perl)`. The input variable for this module is `PERL_EXECUTABLE`.

JackBoosY replied:
Since there is no judgement in
https://github.com/litespeedtech/lsquic/blob/a74702c630e108125e71898398737baec8f02238/CMakeLists.txt#L135

We should not do this unless the following modifications are made:
```patch
+if (NOT PERL_EXECUTABLE)
find_package(Perl)
IF(NOT PERL_FOUND)
    MESSAGE(FATAL_ERROR "Perl not found -- need it to generate source code")
ENDIF()
+endif()
```

dg0yt replied:
@JackBoosY IMO your proposed modification is wrong. The message says that Perl is required. We have to provide Perl via `PERL_EXECUTABLE`. That's all.

JackBoosY replied:
> We have to provide Perl via PERL_EXECUTABLE.

Any evidence?

dg0yt replied:
Perl is not optional because a) upstream says so and b) it is used.

The full story is that we need two definitions, due to an inconsistency in upstream's build system:
- `PERL_EXECUTABLE` to satisfy `find_package(Perl)`.
- `PERL` because this is the variables still used here: https://github.com/litespeedtech/lsquic/blob/0a4f8953dc92dd3085e48ed90f293f052cff8427/src/liblsquic/CMakeLists.txt#L109-L114

Vcpkg maintainer guidelines require that we don' patch if we can provide configuration.

rpavlik replied:
thanks for catching the remaining usage of `PERL` instead of `PERL_EXECUTABLE`, I'll have to upstream that fix.

dg0yt replied:
If you contact upstream, you may also suggest to add `REQUIRED` to `find_package(Perl)`. This removes the necessity for the error handling.

rpavlik replied:
Oh right, not sure why I didn't do that in my previous pr. Good catch.

rpavlik commented:
OK, x64-windows and x64-windows-static confirmed to work. Trying without the "supports" and seeing if 32 bit works with the update to boringssl.

rpavlik commented:
Looks like it's still 64-bit only:

```
C:\vcpkg\buildtrees\liblsquic\src\v3.1.1-4d7de5450b\src\liblsquic\lsquic_rechist.c(321): warning C4389: '==': signed/unsigned mismatch
```

rpavlik commented:
OK, works on x64-windows, x64-windows-static, and x64-linux. This is ready to review and merge. I'm following up on the 32-bit stuff (and the one required cast) and the PERL/PERL_EXECUTABLE

JackBoosY commented:
Testing this port locally.

JackBoosY commented:
Already tested successfully on x64-windows, x64-windows-static, x64-linux.

rpavlik committed:
version

vicroms committed:
Merge branch 'master' of github.com:microsoft/vcpkg into lsquic

vicroms committed:
use vcpkg_install_copyright()
